hera_cal/abscal.py

```diff
@@ -3693,6 +3694,219 @@ def post_redcal_abscal_run(data_file, redcal_file, model_files, raw_auto_file=No
+        which is memory inefficient for analyses using a small fraction of the array
+        but will break if the redundant baselines are keyed to antennas that are not
+        present in the data so only use this if you are confident that this is the case.
+        default is False.
+    clobber: bool, optional
+        overwrite outputs if True.
+    tol: float, optional
+        perform calibration loop until differences in gain solutions are at this level.
+    max_iter: int, optional
+        maximum number of iterations to perform
+        default is 10
+    refant: tuple, optional
+        referance antenna in form of (antnum, polstr).
+        Default is None -> automatically select a refant.
+    ant_threshold: float, optional
+        threshold of flags in frequency and time for a given antenna to completely
+        flag this antenna rom calibration.
+    no_ampcal: bool, optional
+        skip amplitude calibration (only calibrate phases)
+        default is False.
+    no_phscal: bool, optional
+        skip phase calibration (only calibrate amplitudes)
+        default is False.
+    dly_lincal: bool, optional
+        perform initial delay calibration before abscal
+        default is False.
+    verbose: bool, optional
+        lots of outputs.
+    """
+    hdd = io.HERAData(data_file)
+    hdm = io.HERAData(model_file)
+
+    data, data_flags, data_nsamples = hdd.read()
+
+    if precalibration_gain_file is not None:
+        uvc_precal = io.HERACal(precalibration_gain_file)
+        gains_precal, flags_precal, _, _ = uvc_precal.read()
+        # apply precal gains to data
+        calibrate_in_place(data=data, data_flags=data_flags,
+                           new_gains=gains_precal, cal_flags=flags_precal)
+
+    # expand hdm by redundancy
+    hdm.read()
+    if inflate_model_by_redundancy:
+        if constrain_model_to_data_ants:
+            # In order to avoid inflating to full interferometer dataset
+            # which will be unecessary for many earlier analyses
+            # prune model antennas to only include data antennas in hdd
+            # this will only work if the antennas in the baseline keys for each redundant group
+            # in the model are also present in the dataset so only use this if you are confident
+            # that this is the case.
+            all_data_ants = np.unique(np.hstack([hdd.ant_1_array, hdd.ant_2_array]))
+            all_model_ants = np.unique(np.hstack([hdm.ant_1_array, hdm.ant_2_array]))
+            assert np.all([ant in all_data_ants for ant in all_model_ants]), "All model antennas must be present in data if constrain_model_to_data_ants is True!"
+            hdm.select(antenna_nums=all_data_ants,
+                       keep_all_metadata=False)
+        hdm.inflate_by_redundancy()
+        # also make sure to only include baselines present in hdd.
+        hdm.select(bls=hdd.bls)
+
+    model, model_flags, model_nsamples = hdm.read()
```

Comment:
I actually am not sure if this above conditionals are respected when you make a `hdm.read()` call. As i recall from a few years back (this may have changed so @jsdillon please weigh in here), when you do a `hdm.read()` it actually re-loades the filename stored in the object, so all previous antenna / pol selections are actually overwritten.

In order to keep the selections you've done above, you need to use the `_determine_blt_slice` `_determine_pol_slice` and then call `build_datacontainers` directly, which is what the `hdm.read()` call does at the end after re-loading in the file metadata.

Reply:
perhaps a test should be constructed to ensure this is working properly too

Reply:
I've changed the `model, model_flags, model_nsamples = hdm.read()` to `model, model_flags, model_nsamples = hdm.build_datacontainers()`